featurebyte/query_graph/sql/interpreter/tile.py

```diff
@@ -99,25 +104,69 @@ def construct_tile_gen_sql(self, starting_node: Node) -> list[TileGenSql]:
         list[TileGenSql]
         """
         # Groupby operations requires building tiles (assuming the aggregation type supports tiling)
-        tile_generating_nodes = {}
-        for node in self.query_graph.iterate_nodes(starting_node, NodeType.GROUPBY):
-            assert isinstance(node, GroupByNode)
-            tile_generating_nodes[node.name] = node
-
         sqls = []
-        for node in tile_generating_nodes.values():
-            info = self.make_one_tile_sql(node)
+        for groupby_node in self.query_graph.iterate_nodes(starting_node, NodeType.GROUPBY):
+            assert isinstance(groupby_node, GroupByNode)
+
+            if self.is_on_demand:
+                event_table_timestamp_filter = None
+            else:
+                event_table_timestamp_filter = self._get_event_table_timestamp_filter(groupby_node)
+
+            info = self.make_one_tile_sql(groupby_node, event_table_timestamp_filter)
             sqls.append(info)
 
         return sqls
 
-    def make_one_tile_sql(self, groupby_node: GroupByNode) -> TileGenSql:
+    def _get_event_table_timestamp_filter(
+        self, groupby_node: GroupByNode
+    ) -> Optional[EventTableTimestampFilter]:
+        pruned_graph, node_name_map, _ = prune_query_graph(
+            graph=self.query_graph, node=groupby_node
+        )
+
+        # Check there is no lag operation, otherwise cannot apply timestamp filter directly
+        pruned_node = pruned_graph.get_node_by_name(node_name_map[groupby_node.name])
+        for node in dfs_traversal(pruned_graph, pruned_node):
+            if node.type == NodeType.LAG:
+                return None
+
+        # Identify which EventTable can be filtered. This is based on the timestamp column used in
+        # the groupby node.
+        groupby_input_node = pruned_graph.get_node_by_name(
+            pruned_graph.get_input_node_names(pruned_node)[0]
+        )
+        op_struct_info = OperationStructureExtractor(graph=self.query_graph).extract(
+            groupby_input_node
+        )
+        op_struct = op_struct_info.operation_structure_map[groupby_input_node.name]
+        for column in op_struct.columns:
+            if (
+                column.name == groupby_node.parameters.timestamp
+                and isinstance(column, SourceDataColumn)
+                and column.table_id is not None
+            ):
+                assert isinstance(column, SourceDataColumn)
```

Comment:
is this assert necessary since it's already in the if-statement?

Reply:
Yes, this is necessary to make mypy happy. I think not necessary when the if condition is simpler.